### PR TITLE
[SPARK-38974][SQL] Filter registered functions with a given database name in list functions

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR fixes a bug in list functions to filter out registered functions that do not belong to the specified database.

### Why are the changes needed?

To fix a bug for `SHOW FUNCTIONS IN [db]`. Listed functions should only include all temporary functions and persistent functions in the specified database, instead of all registered functions in the function registry.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test

### CODE REVIEW

**Comment by allisonwang-db** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 1]
```diff
@@ -1753,6 +1753,23 @@ class SessionCatalog(
     }
   }
 
+  /**
+   * List all registered functions in a database with the given pattern.
+   */
+  private def listRegisteredFunctions(db: String, pattern: String): Seq[FunctionIdentifier] = {
+    val functions = (functionRegistry.listFunction() ++ tableFunctionRegistry.listFunction())
+      .filter(_.database.forall(_ == db))
+    StringUtils.filterPattern(functions.map(_.unquotedString), pattern).map { f =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1762]: I think the pattern is to match the function name, not the qualified function identifier.
```suggestion
    StringUtils.filterPattern(functions.map(_.name), pattern).map { f =>
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1762]: Ah, we probably need to add a variant of `StringUtils.filterPattern` which returns matching indexes.

Comment (allisonwang-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1762]: https://github.com/apache/spark/pull/35056#issue-1090368357 if we still want to support the legacy `SHOW FUNCTIONS LIKE 'db.func'`, then we probably can't filter out the registered functions from another database. 

But `SHOW FUNCTIONS LIKE 'db.func'` can only show **registered** functions from another database, and is different from `SHOW FUNCTIONS IN db LIKE 'func'`.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3!

**Pull request closed without merging**